### PR TITLE
feat(jax): force & virial

### PR DESCRIPTION
**Pull request by njzjz**
## Release Notes

- **New Features**
	- Introduced new methods `forward_common_atomic` in multiple classes to enhance atomic model predictions and derivative calculations.
	- Added a new function `get_leading_dims` for better handling of output dimensions.
	- Added a new function `scatter_sum` for performing reduction operations on tensors.
	- Updated test methods to include flexible handling of results with the new `SKIP_FLAG` variable.

- **Bug Fixes**
	- Improved numerical stability in calculations by ensuring small values are handled appropriately.

- **Tests**
	- Expanded test outputs to include additional data like forces and virials for more comprehensive testing.
	- Enhanced backend handling in tests to accommodate new return values based on backend availability.

### CODE REVIEW

File: deepmd/dpmodel/model/transform_output.py [hunk 1]
```diff
@@ -47,6 +48,15 @@ def fit_output_to_model_output(
     return model_ret
 
 
+def get_leading_dims(
+    vv: np.ndarray,
+    vdef: OutputVariableDef,
+):
+    """Get the dimensions of nf x nloc."""
+    vshape = vv.shape
+    return list(vshape[: (len(vshape) - len(vdef.shape))])
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/model/transform_output.py:57]: Generate numpy-style doc

[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/model/transform_output.py]: added in https://github.com/deepmodeling/deepmd-kit/pull/4251/commits/004b89a886a26b49dcef8b0a298cd45532537a07

File: deepmd/dpmodel/model/transform_output.py [hunk 2]
```diff
@@ -65,15 +89,68 @@ def communicate_extended_output(
...
+                        v_idx = xp.arange(virial.size, dtype=xp.int64).reshape(
+                            virial.shape
+                        )
+                        new_idx = jnp.take_along_axis(v_idx, mapping, axis=1).ravel()
+                        v_shape = virial.shape
+                        virial = virial.ravel()
+                        virial = virial.at[new_idx].add(model_ret[kk_derv_c].ravel())
+                        virial = virial.reshape(v_shape)
+                    else:
+                        raise NotImplementedError("Only JAX arrays are supported.")
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/dpmodel/model/transform_output.py]: Fixed in https://github.com/deepmodeling/deepmd-kit/pull/4251/commits/b9eefd3afd8e9728b95dda9abb38702ea033a88b

File: deepmd/jax/model/base_model.py [hunk 3]
```diff
@@ -1,6 +1,101 @@
...
+                    ):
+                        atomic_ret = self.atomic_model.forward_common_atomic(
+                            cc_ext[None, ...],
+                            extended_atype[None, ...],
+                            nlist[None, ...],
+                            mapping=mapping[None, ...] if mapping is not None else None,
+                            fparam=fparam[None, ...] if fparam is not None else None,
+                            aparam=aparam[None, ...] if aparam is not None else None,
+                        )
+                        return jnp.sum(atomic_ret[kk][0], axis=atom_axis)[ss]
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/jax/model/base_model.py]: Fixed in https://github.com/deepmodeling/deepmd-kit/pull/4251/commits/b9eefd3afd8e9728b95dda9abb38702ea033a88b

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/jax/model/base_model.py [hunk 4]
```diff
@@ -1,6 +1,111 @@
...
+            kk_redu = get_reduce_name(kk)
+            model_predict[kk_redu] = jnp.sum(vv, axis=atom_axis)
+            kk_derv_r, kk_derv_c = get_deriv_name(kk)
+            if vdef.c_differentiable:
+                size = 1
+                for ii in vdef.shape:
+                    size *= ii
+
+                split_ff = []
+                split_vv = []
```
Comment (wanghan-iapcm) [inline deepmd/jax/model/base_model.py:56]: I do not think we need to split and then take derivative. 
you may use [`jax.jacrev`](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev) for computing the force without splitting the output.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/jax/model/base_model.py (hunk 4 above)
Comment (njzjz) [reply deepmd/jax/model/base_model.py:56]: It's not a problem in theory, but handling the shapes for the following steps is required...

https://github.com/deepmodeling/deepmd-kit/blob/23e318a61a19f3618070bc80e89e0209796d4513/deepmd/jax/model/base_model.py#L89-L90

Comment (njzjz) [reply deepmd/jax/model/base_model.py]: Please check my inline comments in 19e09708d679f12460218b0546fdf558de6cb68a

**Review by wanghan-iapcm — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

**Pull request merged**